server/src/main/java/org/apache/druid/metadata/SQLMetadataSegmentManager.java

```diff
@@ -102,9 +104,14 @@
   private final Supplier<MetadataStorageTablesConfig> dbTables;
   private final SQLMetadataConnector connector;
 
-  private ConcurrentHashMap<String, DruidDataSource> dataSources = new ConcurrentHashMap<>();
+  // Volatile since this reference is reassigned in "poll" and then read from in other threads.
```

Comment:
This comment doesn't explain *why* does the field need to be `volatile`. The underlying reason is that the field is effectively a lazily initialized field, and the absence of `volatile` may lead to NPE unless the rest of the code always reads the field to local variables before using, that is too much of a burden for developers: https://github.com/code-review-checklists/java-concurrency#safe-local-dcl

(Actually, as you translated all code to monadic Optional use of `dataSources` with a single read, it *does not* need to be volatile, but I would say that those monadic Optional chains are worse than simple if-else.)

Reply:
Ok, because of this: https://github.com/apache/incubator-druid/pull/7447#discussion_r274563782 the previous comment is irrelevant, there is actually no reason why the field should be `volatile` in the current version of the code.

Reply:
Are you saying it is fine to have a field that is written from one thread, and read from another, with no synchronization or volatile marker, as long as each reader reads it into a local variable first? My understanding of the JMM is that in this case there's no happens-before relationship established, and all bets are off - readers have no guarantees around _ever_ reading anything nonnull (although in practice they probably will, but that's not something you'd want to depend on).

Reply:
Practically, as you noted, it doesn't matter (on x86 platform which Druid targets). Formally, `volatile` is still not enough to ensure "ever reading non-null" before Java 9 where it was formalized [in this document](http://gee.cs.oswego.edu/dl/html/j9mm.html).